### PR TITLE
executor, expression: fix question that sql results of function REGEXP are inconsistent with MySQL

### LINKED ISSUE #20060: These sql results of function REGEXP are inconsistent with MySQL
## Bug Report

These sql results are inconsistent with MySQL

### 1. Minimal reproduce step (Required)
MySQL 5.7：
```mysql
mysql> SELECT 'new*\n*line' REGEXP 'new\\*.\\*line';
+---------------------------------------+
| 'new*\n*line' REGEXP 'new\\*.\\*line' |
+---------------------------------------+
|                                     1 |
+---------------------------------------+
1 row in set (0.01 sec)

mysql> SELECT '~' REGEXP '[[.~.]]';
+----------------------+
| '~' REGEXP '[[.~.]]' |
+----------------------+
|                    1 |
+----------------------+
1 row in set (0.00 sec)

mysql> SELECT '~' REGEXP '[[.tilde.]]';
+--------------------------+
| '~' REGEXP '[[.tilde.]]' |
+--------------------------+
|                        1 |
+--------------------------+
1 row in set (0.00 sec)
```

MySQL 8.0:
```mysql
mysql> SELECT 'new*\n*line' REGEXP 'new\\*.\\*line';
+---------------------------------------+
| 'new*\n*line' REGEXP 'new\\*.\\*line' |
+---------------------------------------+
|                                     0 |
+---------------------------------------+
1 row in set (0.00 sec)

mysql> SELECT '~' REGEXP '[[.~.]]';
+----------------------+
| '~' REGEXP '[[.~.]]' |
+----------------------+
|                    1 |
+----------------------+
1 row in set (0.00 sec)

mysql> SELECT '~' REGEXP '[[.tilde.]]';
+--------------------------+
| '~' REGEXP '[[.tilde.]]' |
+--------------------------+
|                        0 |
+--------------------------+
1 row in set (0.00 sec)

mysql> select version();
+-----------+
| version() |
+-----------+
| 8.0.21    |
+-----------+
1 row in set (0.00 sec)
```

TiDB：
```mysql
MySQL [db1]> SELECT 'new*\n*line' REGEXP 'new\\*.\\*line';
+---------------------------------------+
| 'new*\n*line' REGEXP 'new\\*.\\*line' |
+---------------------------------------+
|                                     0 |
+---------------------------------------+
1 row in set (0.00 sec)

MySQL [db1]> SELECT '~' REGEXP '[[.~.]]';
+----------------------+
| '~' REGEXP '[[.~.]]' |
+----------------------+
|                    0 |
+----------------------+
1 row in set (0.00 sec)

MySQL [db1]> SELECT '~' REGEXP '[[.tilde.]]';
+--------------------------+
| '~' REGEXP '[[.tilde.]]' |
+--------------------------+
|                        0 |
+--------------------------+
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)
```
| Release Version: v4.0.3
Edition: Community
Git Commit Hash: 0529b1b493e46aae71bbe34cbe24515a2eb1b47c
Git Branch: heads/refs/tags/v4.0.3
UTC Build Time: 2020-07-24 12:06:35
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

## SIG slack channel
[#sig-exec](https://slack.tidb.io/invite?team=tidb-community&channel=sig-exec&ref=high-performance)

## Score
- 300

## Mentor
 * @qw4990

### PR DESCRIPTION
**Pull request by xiaodong-ji**
### What problem does this PR solve?

Issue Number: close #20060

Problem Summary: These sql results of function REGEXP are inconsistent with MySQL.

### What is changed and how it works?

What's Changed: If there are double '[' or ']' in the pattern, eg: '[[.~.]]', remove one of them.

### Check List

Tests

- Unit test

### Release note

- fix question that sql results of function REGEXP are inconsistent with MySQL

### CODE REVIEW

**Comment by wjhuang2016** [comment]
> If there are double '[' or ']' in the pattern, eg: '[[.~.]]', remove one of them.

@xiaodong-ji Hi, thank you for your contribution, could you provide some information about this change?
I mean, though MySQL's result looks like that it's true, we have better confirm it from the document or source code.

**Comment by xiaodong-ji** [comment]
> > If there are double '[' or ']' in the pattern, eg: '[[.~.]]', remove one of them.
> 
> @xiaodong-ji Hi, thank you for your contribution, could you provide some information about this change?
> I mean, though MySQL's result looks like that it's true, we have better confirm it from the document or source code.

I seem to have only found this information (` [[=o=]], [[=(+)=]], and [o(+)] are all synonymous`) in the follow document.

**Review by time-and-fate — COMMENTED** [review]
Thanks for your contributions.
Seems like your change will make `SELECT 'z' REGEXP '[[:alpha:]]';` return 0, which currently is and should be 1.
I think simply change `[[...]]` to `[...]` isn't a proper method to solve the issue.

**Comment by xiaodong-ji** [comment]
> Thanks for your contributions.
> Seems like your change will make `SELECT 'z' REGEXP '[[:alpha:]]';` return 0, which currently is and should be 1.
> I think simply change `[[...]]` to `[...]` isn't a proper method to solve the issue.

You are right, is there any ideas to slove this problem.

File: expression/builtin_like.go [hunk 1]
```diff
@@ -161,6 +161,26 @@ func (b *builtinRegexpSharedSig) evalInt(row chunk.Row) (int64, bool, error) {
...
+		}
+	}
+	if len(l) > 0 && len(r) > 0 && len(l) == len(r) {
+		for i := 0; i < len(r); i++ {
+			patBytes = append(patBytes[:l[i]], patBytes[l[i]+1:]...)
+			patBytes = append(patBytes[:r[i]], patBytes[r[i]+1:]...)
+		}
+		pat = string(patBytes)
+	}
+
```
Comment (rebelice) [inline expression/builtin_like.go:183]: It seems that your solution can't handle cases like `'\\[[\\]]'`.

**Review by rebelice — CHANGES_REQUESTED** [review]

**Review by wjhuang2016 — COMMENTED** [review]
In fact, it's very difficult to make REGEXP to be compatible with MySQL. Because the regexp library is different, and we need to consider Go's, Rust's, and C++'s library. I think we can ignore this compatible problem so far. But if you have any good idea, feel free to discuss with us.

**Comment by wjhuang2016** [comment]
@xiaodong-ji Hi, as I said before, it's very difficult to make REGEXP to be compatible with MySQL. And this PR's modification seems dangerous. We decide not to merge this PR, sorry for that.

**Comment by xiaodong-ji** [comment]
> @xiaodong-ji Hi, as I said before, it's very difficult to make REGEXP to be compatible with MySQL. And this PR's modification seems dangerous. We decide not to merge this PR, sorry for that.

OK，it doesn't matter.

**Pull request closed without merging**